designs/function-node-lifecycle/README.md

```diff
@@ -41,17 +41,120 @@ Following figure shows new lifecycle model of function node.  While creating a f
+| denyList | array of string | no | List of regular expressions that matches module's NPM registry specification. External libraries that matches this list **can not** be used in `Function` nodes. |
+
+`allowList` precedes over `denyList`.
+
+```
+    // Example of settings.js
+    externalModules: {
+        mode: "manual",
+        allowList: [
+            "^fs-ext$",       // allow fs-ext
+            "^qrcode@1.4.4$"  // allow qrcode version 1.4.4
+        ],
+        denyList: [
+            ".*"              // disallow others 
+        ]
+    },
+```
+If the `externalModule` property not exists, `none` mode and denyList of `".*"`, meaning not allow installation and use of external module, is expected as a default.
+
+#### New API for installing NPM module
+
+In order to allow NPM modules to be installed automatically before node execution, `RED.nodes.registerType` will be extended to accept new `dynamicModuleList` property in an optional third parameter.  It may have following value:
+
+1.  a `string` value that represents a property name of a node instance.  The property of the node instance must have a list of modules to install.
+
+   ```
+   // Node definition
+   RED.nodes.registerType("function", FunctionNode, {
+       dynamicModuleList: "modules"
+   });
+   ```
+
+   ```
+   // Node instance definition
+   {
+       ...
+       "modules": [ "fs-ext@1.2.3", "qrcode@1.4.4" ]
+       ...
+   }
+   ```
+
+2. or an array of objects that must have a `name` property and may have a `scope` propety (but can have any other properties)
+
+   ```
+   // Node definition
+   RED.nodes.registerType("function", FunctionNode, {
+       dynamicModuleList: "modules"
+   });
+   ```
+
+   ```
+   // Node instance definition
+   {
+       ...
+       "modules": [ 
+           { 
+               "name": "fs-ext@1.2.3", 
+               "scope": "function:xyz",
+               "var": "fs", 
+               ... 
+           },
+           { 
+               "name": "qrcode@1.4.4", 
+               "scope": "function:xyz",
+               "var": "qrcode", 
+               ... 
+           }
+       ],
+       ...
+   }
+   ```
+
+   The Node-RED runtime will scan the list of modules to install in nodes deploy process before starting nodes execution.  A module in the list will be installed, if auto mode is specified in `settings.js` .  If `scope` is specified, NPM module is installed locally to specified scope.  On removal of a node with locally installed NPM modules, the node must uninstall the installed modules.
+
+   In order to support installation of NPM modules, the Node-RED runtime provides following API:
+
+   - Runtime API for Nodes
+
+     - `RED.nodes.addModule(opts)`
+     - `RED.nodes.removeModule(opts)`
+     - `RED.nodes.getModuleInfo(opts)`
+
+     Node API for nodes is extended to include these functions.  These new APIs accept the same parameter object `opts` to corresponding runtime APIs but works on generic NPM modules. 
```

Comment:
> Node API for nodes is extended

Are you referring to the `RED` object passed to nodes? I don't think we should expose them to the node - there should be no reason for a node to add/remove modules of its own accord.

The idea is for the `dynamicModuleList` property to be used by the runtime to identify what modules are needed on behalf of the individual nodes. Nodes shouldn't be given the freedom to dynamically install/remove nodes themselves.

Reply:
I may not understand your point correctly.
But, these APIs are intended for supporting interactive NPM module installation panel as described in
https://github.com/node-red/designs/pull/31#discussion_r537273071